File: Dart/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCycle.dart
import 'dart:collection';

class ListNode {
  int val;
  ListNode? next;
  ListNode(this.val, [this.next]);
}

void main() {
  // Creating a linked list with a cycle: 3 -> 2 -> 0 -> -4 -> 2 (cycle back to 2)
  ListNode head = ListNode(3);
  head.next = ListNode(2);
  head.next!.next = ListNode(0);
  head.next!.next!.next = ListNode(-4, head.next); // Cycle here, -4 points back to 2

  // Testing different cycle detection algorithms
  print("Detect Cycle (Floyd's Cycle Detection): ${hasCycleFloyd(head)}");
  print("Detect Cycle (HashSet): ${hasCycleHashSet(head)}");
  print("Detect Cycle (Two-Pointer Optimization): ${hasCycleTwoPointer(head)}");
  print("Detect Cycle (Node Marking): ${hasCycleMarking(head)}");
  print("Detect Cycle (Length Calculation Approach): ${hasCycleLength(head)}");
}

// Approach 1: Floyd’s Cycle Detection (Tortoise and Hare Algorithm)
// Explanation:
// - Uses two pointers, 'slow' and 'fast'.
// - 'slow' moves one step at a time, 'fast' moves two steps.
// - If there's a cycle, 'fast' will eventually catch up to 'slow'.
// - If 'fast' reaches null, there's no cycle.
bool hasCycleFloyd(ListNode? head) {
  ListNode? slow = head, fast = head;
  while (fast != null && fast.next != null) {
    slow = slow?.next; // Move slow one step
    fast = fast.next?.next; // Move fast two steps
    if (slow == fast) return true; // Cycle detected
  }
  return false; // No cycle found
}

// Approach 2: Using HashSet to Store Visited Nodes
// Explanation:
// - Uses a HashSet to keep track of visited nodes.
// - If a node is visited again, it means there's a cycle.
// - Uses extra space for the HashSet.
bool hasCycleHashSet(ListNode? head) {
  HashSet<ListNode> visited = HashSet<ListNode>();
  while (head != null) {
    if (visited.contains(head)) return true; // Cycle detected
    visited.add(head); // Mark node as visited
    head = head.next;
  }
  return false; // No cycle found
}

// Approach 3: Two-Pointer Optimization (Check for nulls early)
// Explanation:
// - Similar to Floyd's algorithm, but checks for nulls at the beginning.
// - Slightly optimizes the code by preventing unnecessary null checks inside the loop.
bool hasCycleTwoPointer(ListNode? head) {
  if (head == null || head.next == null) return false; // Early null check
  ListNode? slow = head, fast = head.next; // fast starts one step ahead
  while (fast != null && fast.next != null) {
    if (slow == fast) return true; // Cycle detected
    slow = slow?.next;
    fast = fast.next?.next;
  }
  return false; // No cycle found
}

// Approach 4: Marking Nodes with a Special Value
// Explanation:
// - Modifies the 'val' of visited nodes to a special marker value (999999).
// - If a node with the marker value is encountered, it means there's a cycle.
// - Modifies the original linked list.
bool hasCycleMarking(ListNode? head) {
  while (head != null) {
    if (head.val == 999999) return true; // Cycle detected
    head.val = 999999; // Mark node as visited
    head = head.next;
  }
  return false; // No cycle found
}

// Approach 5: Detect Cycle by Finding Length of Cycle
// Explanation:
// - Uses Floyd's algorithm to detect the cycle.
// - If a cycle is detected, it calculates the length of the cycle.
// - Useful if you need to know the length of the cycle.
bool hasCycleLength(ListNode? head) {
  ListNode? slow = head, fast = head;
  while (fast != null && fast.next != null) {
    slow = slow?.next;
    fast = fast.next?.next;
    if (slow == fast) {
      // Cycle detected, calculate cycle length
      int length = 1;
      ListNode? temp = slow?.next;
      while (temp != null && temp != slow) {
        length++;
        temp = temp.next;
      }
      print("Cycle Length: $length"); // Print the cycle length
      return true;
    }
  }
  return false; // No cycle found
}

File: Java/6-LinkedList/Problems/3-LinkedListCycle/LinkedListCycle.java
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;

public class LinkedListCycle {

    // Definition of a singly linked list node.
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public static void main(String[] args) {
        // Create a sample linked list with a cycle for testing.
        ListNode head = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);

        head.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node2; // Creating a cycle: node5 points back to node2.

        // Test the different cycle detection methods.
        System.out.println("Using HashSet: Cycle detected? " + hasCycleUsingHashSet(head)); // Expected: true
        System.out.println("Using Floyd's Tortoise and Hare: Cycle detected? " + hasCycleUsingFloyd(head)); // Expected: true
        System.out.println("Using Stack: Cycle detected? " + hasCycleUsingStack(head)); //Expected: true
        System.out.println("Using Visited Flag: Cycle detected? " + hasCycleUsingVisitedFlag(head)); //Expected: true
        System.out.println("Using Length Comparison: Cycle detected? " + hasCycleUsingLengthComparison(head)); //Expected: true

        // Create a sample linked list without a cycle
        ListNode head2 = new ListNode(1);
        ListNode node6 = new ListNode(2);
        ListNode node7 = new ListNode(3);
        head2.next = node6;
        node6.next = node7;

        System.out.println("Using HashSet: Cycle detected? " + hasCycleUsingHashSet(head2)); // Expected: false
        System.out.println("Using Floyd's Tortoise and Hare: Cycle detected? " + hasCycleUsingFloyd(head2)); // Expected: false
        System.out.println("Using Stack: Cycle detected? " + hasCycleUsingStack(head2)); //Expected: false
        System.out.println("Using Visited Flag: Cycle detected? " + hasCycleUsingVisitedFlag(head2)); //Expected: false
        System.out.println("Using Length Comparison: Cycle detected? " + hasCycleUsingLengthComparison(head2)); //Expected: false
    }

    // 1. Using HashSet
    // Time Complexity: O(n), Space Complexity: O(n)
    // Description: This method uses a HashSet to keep track of visited nodes.  If a node is visited again,
    //              it means there is a cycle.
    public static boolean hasCycleUsingHashSet(ListNode head) {
        if (head == null) {
            return false;
        }
        Set<ListNode> visitedNodes = new HashSet<>();
        ListNode current = head;
        while (current != null) {
            if (visitedNodes.contains(current)) {
                return true; // Cycle detected: current node has been visited before.
            }
            visitedNodes.add(current); // Add the current node to the visited set.
            current = current.next;     // Move to the next node.
        }
        return false; // No cycle found: the list ends without repeating a node.
    }

    // 2. Floyd's Tortoise and Hare (Fast and Slow Pointer)
    // Time Complexity: O(n), Space Complexity: O(1)
    // Description: This is the most efficient method. It uses two pointers, one moving slowly (tortoise) and the other
    //              moving faster (hare). If there is a cycle, the two pointers will eventually meet.
    public static boolean hasCycleUsingFloyd(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;       // Tortoise moves one step at a time.
            fast = fast.next.next; // Hare moves two steps at a time.
            if (slow == fast) {
                return true; // Cycle detected: the pointers meet.
            }
        }
        return false; // No cycle found: the fast pointer reached the end of the list.
    }

    // 3. Using Stack
    // Time Complexity: O(n), Space Complexity: O(n)
    // Description: This method uses a stack to store the visited nodes.  If a node is encountered that is already in the stack,
    //              then a cycle exists.
    public static boolean hasCycleUsingStack(ListNode head) {
        if (head == null) {
            return false;
        }
        Stack<ListNode> stack = new Stack<>();
        ListNode current = head;
        while (current != null) {
            if (stack.contains(current)) {
                return true; // Cycle detected
            }
            stack.push(current);
            current = current.next;
        }
        return false; // No cycle
    }

    // 4. Using Visited Flag (Modifying the Linked List)
    // Time Complexity: O(n), Space Complexity: O(1)
    // Description: This method modifies the linked list by adding a "visited" flag to each node.
    //              If a node is visited again, it indicates a cycle.  This method modifies the original list.
    public static boolean hasCycleUsingVisitedFlag(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode current = head;
        while (current != null) {
            if (current.val == Integer.MIN_VALUE) {
                return true; // Cycle detected: node already marked as visited.
            }
            current.val = Integer.MIN_VALUE; // Mark the current node as visited.
            current = current.next;
        }
        return false; // No cycle found.
    }

    // 5. Using Length Comparison
    // Time Complexity: O(n^2), Space Complexity: O(1)
    // Description: This method calculates the length of the linked list by traversing it. In each iteration, it checks
    //              if the length of the remaining list is decreasing. If not, it implies a cycle.
    public static boolean hasCycleUsingLengthComparison(ListNode head) {
        if (head == null) {
            return false;
        }
        ListNode current = head;
        int originalLength = 0;
        while(current != null){
            originalLength++;
            current = current.next;
        }

        current = head;
        int currentLength = originalLength;
        while (current != null) {
            int tempLength = 0;
            ListNode temp = current;
            while(temp != null){
                tempLength++;
                temp = temp.next;
            }
            if (tempLength > currentLength) {
                return true;
            }
            currentLength = tempLength;
            current = current.next;
        }
        return false;
    }
}

